### DISCUSSION TITLE
IAsyncEnumerable without Await

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4062] [Q&A] [answered]
What is the right way to implement an IAsyncEnumerable that doesn't actually await anything so that you don't get any warnings?

**Comment by CyrusNajmabadi** [comment]
can you provide more details (including code)?  I don't know what warnings you're referring to.

**Reply by TonyValenti** [reply]
Here is sample code:
```
    public abstract class Foo {
        public abstract IAsyncEnumerable<int> DoWorkAsync();
    }

    public abstract class Foo2 : Foo {
        public override async IAsyncEnumerable<int> DoWorkAsync() {
            yield return 1;
        }
    }
```

It gives This Async Method Lacks Await...". 

Removing the "Async" gives "Must be Async to return IAsyncEnumerable".

How would I implement a method like that without any warnings?

**Reply by CyrusNajmabadi** [reply]
```c#
public abstract class Foo2 : Foo
{
#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
    public override async IAsyncEnumerable<int> DoWorkAsync()
#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
    {
        yield return 1;
    }
}
```

**Reply by TonyValenti** [reply]
Is there a way to do it without disabling the warning?

**Reply by TonyValenti** [reply]
I tried ```yield return Task.FromResult(1)``` and that didnt' do it either.

**Reply by RikkiGibson** [reply]
ping @jcouv in case you have any suggestions for this scenario :)

**Reply by colejohnson66** [reply]
Can you just remove the `async` qualifier?

**Comment by jcouv** [comment, marked as answer]
Options I can think of:
1. Cheapest way is to include an `await Task.CompletedTask;` in your code
2. Implementing an adapter type by hand (your own types implementing `IAsyncEnumerable` and `IAsyncEnumerator`, which would wrap around an `IEnumerator`), which is a bit tedious
3. Find a library that already implemented such wrapper types
4. Raise the question on StackOverflow and see if any better ideas come up ;-)

**Comment by Jason5Lee** [comment]
You can add `await Task.Yield()` before `yield return` .

**Reply by HaloFour** [reply]
`await Task.CompletedTask;` would be more efficient since the awaiter is already completed so the method will continue synchronously.  `Task.Yield()` yields the thread and posts the continuation to the synchronization context or thread pool.

But yeah, the theory is the same.  Awaiting on anything will silence the warning.  It would be nice if there was a helper for this, like a way to wrap a scalar `T` into an `IAsyncEnumerable<T>`.

**Reply by Richiban** [reply]
Like a `Task.FromResults`, for example? A naive implementation might look like:

```csharp
    public async IAsyncEnumerable<T> FromResults<T>(params T[] items)
    {
        await Task.CompletedTask;
        
        foreach (var item in items)
        {
            yield return item;
        }
    }
```

**Reply by silkfire** [reply]
You can put the statement at the end of your method:

```csharp
public async IAsyncEnumerable<T> FromResults<T>(params T[] items)
{
   foreach (var item in items)
   {
      yield return item;
   }

   await Task.CompletedTask;
}
```
